genai_processors/contrib/openrouter_model.py

```diff
@@ -0,0 +1,424 @@
+"""Wraps the OpenRouter API into a Processor.
+
+This module provides access to hundreds of AI models through OpenRouter's
+unified API. OpenRouter automatically handles fallbacks and selects the most
+cost-effective options while providing access to models from OpenAI, Anthropic,
+Google, Meta, and many other providers.
+
+## Example usage
+
+```py
+p = OpenRouterModel(
+    api_key="your-openrouter-api-key",
+    model_name="google/gemma-3-27b-it",
+    generate_content_config=GenerateContentConfig(
+        temperature=0.7,
+        max_tokens=1000,
+    )
+)
+```
+
+### Sync Execution
+
+```py
+INPUT_PROMPT = 'Write a haiku about artificial intelligence'
+
+content = processors.apply_sync(p, [ProcessorPart(INPUT_PROMPT)])
+for part in content:
+  if part.text:
+    print(part.text)
+```
+
+### Async Execution
+
+```py
+input_stream = processor.stream_content([ProcessorPart(INPUT_PROMPT)])
+async for part in p(input_stream):
+  if part.text:
+    print(part.text)
+```
+
+### Available Models
+
+OpenRouter provides access to hundreds of models including:
+- OpenAI: gpt-4o, gpt-4-turbo, gpt-3.5-turbo
+- Anthropic: claude-3-5-sonnet, claude-3-opus, claude-3-haiku
+- Google: gemma-3-27b-it, gemini-2.5-pro
+- Meta: llama-3.3-70b-instruct, llama-4-maverick
+- And many more...
+
+For a complete list, visit: https://openrouter.ai/models
+"""
+
+import asyncio
+from collections.abc import AsyncIterable
+import json
+from typing import Any, Literal
+
+from genai_processors import content_api
+from genai_processors import processor
+from google.genai import types as genai_types
+import httpx
+from typing_extensions import TypedDict
+
+
+_DEFAULT_BASE_URL = 'https://openrouter.ai/api/v1'
+_DEFAULT_TIMEOUT = 300
+
+
+class GenerateContentConfig(TypedDict, total=False):
+  """Optional model configuration parameters for OpenRouter."""
+
+  temperature: float | None
+  """Controls randomness in the response. Range: 0.0 to 2.0"""
+
+  top_p: float | None
+  """Controls diversity via nucleus sampling. Range: 0.0 to 1.0"""
+
+  top_k: int | None
+  """Limits the number of tokens to consider for each step."""
+
+  frequency_penalty: float | None
+  """Reduces repetition. Range: -2.0 to 2.0"""
+
+  presence_penalty: float | None
+  """Reduces repetition based on token presence. Range: -2.0 to 2.0"""
+
+  repetition_penalty: float | None
+  """Alternative to frequency_penalty. Range: 0.0 to 2.0"""
+
+  min_p: float | None
+  """Alternative to top_p. Minimum probability threshold."""
+
+  top_a: float | None
+  """Alternative to top_p. Top-a sampling parameter."""
+
+  seed: int | None
+  """Random seed for deterministic outputs."""
+
+  max_tokens: int | None
+  """Maximum number of tokens to generate."""
+
+  response_format: dict[str, Any] | None
+  """Response format specification (e.g., {"type": "json_object"})"""
+
+  stop: list[str] | str | None
+  """Stop sequences to end generation."""
+
+  stream: bool
```

Comment:
Is there a reason to not stream? I suggest to remove this option and always stream.

Non streaming (unary) option is usually there to simplify the API for a common use case, and especially if you deal with raw HTTP. But in Python we can provide better options. E.g. dealing with multimodal content forces us to use `Parts`, so your code has to deal with a list[Part] anyway. And iterable[Part] (which streaming provides) is not that much different.